syntax = "proto3";
package Nrmk.IndyFramework;
import "control_msgs.proto";
import "config_msgs.proto";
import "common_msgs.proto";

service Control {

  rpc GetControlInfo(Empty) returns (ControlInfo) {}
  rpc ActivateIndySDK(SDKLicenseInfo) returns (SDKLicenseResp) {}

  // ---------------------------------------------------------------- //
  // Motion
  // ---------------------------------------------------------------- //
  rpc MoveJ(MoveJReq) returns (Response) {}
  rpc MoveJT(MoveJTReq) returns (Response) {}

  rpc MoveL(MoveLReq) returns (Response) {}
  rpc MoveLT(MoveLTReq) returns (Response) {}

  rpc MoveC(MoveCReq) returns (Response) {}
  rpc MoveCT(MoveCTReq) returns (Response) {}

  rpc MoveJointTraj(MoveJointTrajReq) returns (Response) {}
  rpc MoveTaskTraj(MoveTaskTrajReq) returns (Response) {}

  rpc MoveConveyor(MoveConveyorReq) returns (Response) {}

  rpc WaitIO(WaitIOReq) returns (Response) {}
  rpc WaitTime(WaitTimeReq) returns (Response) {}
  rpc WaitProgress(WaitProgressReq) returns (Response) {}
  rpc WaitTraj(WaitTrajReq) returns (Response) {}
  rpc WaitRadius(WaitRadiusReq) returns (Response) {}
//  rpc MotionStop(MotionStopReq) returns (Response) {}

  rpc MoveJCond(MoveJCondReq) returns (Response) {}

  rpc MoveLinearAxis(MoveAxisReq) returns (Response) {}

  // ---------------------------------------------------------------- //
  // Command
  // ---------------------------------------------------------------- //
  rpc StopMotion(StopCat) returns (Response) {}
  rpc PauseMotion(PauseCat) returns (Response) {}
  rpc SetDirectTeaching(State) returns (Response) {}
  rpc SetSimulationMode(State) returns (Response) {}
  rpc SetCustomControlMode(IntMode) returns (Response) {}  // 0: default controller | 1~: reserved for custom control
  rpc GetCustomControlMode(Empty) returns (IntMode) {}     // 0: default controller | 1~: reserved for custom control
  rpc SetFrictionCompensation(State) returns (Response) {}
  rpc GetFrictionCompensationState(Empty) returns (State) {}
  rpc SetTactTime(TactTime) returns (Response) {}
  rpc GetTactTime(Empty) returns (TactTime) {}

  // ---------------------------------------------------------------- //
  // Violation Recovery
  // ---------------------------------------------------------------- //
  rpc Recover(Empty) returns (Response) {}
  rpc Reboot(Empty) returns (Response) {}
  rpc SetManualRecovery(State) returns (Response) {}
  rpc MoveRecoverJoint(TargetJ) returns (Response) {}

  // ---------------------------------------------------------------- //
  // Program
  // ---------------------------------------------------------------- //
  rpc SearchProgram(Program) returns (ProgramInfo) {}
  rpc PlayProgram(Program) returns (Response) {}
  rpc PauseProgram(Empty) returns (Response) {}
  rpc ResumeProgram(Empty) returns (Response) {}
  rpc StopProgram(Empty) returns (Response) {}
  rpc SendAlarm(Message) returns (Empty) {}
  rpc SendAnnotation(Message) returns (Empty) {}
  rpc PlayTuningProgram(TuningProgram) returns (CollisionThresholds) {}
  rpc PlayProgramLine(Program) returns (Response) {}

  // ---------------------------------------------------------------- //
  // Custom Variable
  // ---------------------------------------------------------------- //
  rpc SetModbusVariableNameList(ModbusVariableList) returns (Empty) {}

  rpc SetVariableNameList(AllVars) returns (Empty) {}
  rpc GetVariableNameList(Empty) returns (AllVars) {}

  rpc SetIntVariable(IntVars) returns (Empty) {}
  rpc GetIntVariable(Empty) returns (IntVars) {}

  rpc SetModbusVariable(ModbusVars) returns (Empty) {}
  rpc GetModbusVariable(Empty) returns (ModbusVars) {}

  rpc SetBoolVariable(BoolVars) returns (Empty) {}
  rpc GetBoolVariable(Empty) returns (BoolVars) {}

  rpc SetFloatVariable(FloatVars) returns (Empty) {}
  rpc GetFloatVariable(Empty) returns (FloatVars) {}

  rpc SetJPosVariable(JPosVars) returns (Empty) {}
  rpc GetJPosVariable(Empty) returns (JPosVars) {}

  rpc SetTPosVariable(TPosVars) returns (Empty) {}
  rpc GetTPosVariable(Empty) returns (TPosVars) {}

  // ---------------------------------------------------------------- //
  // Utility
  // ---------------------------------------------------------------- //
  rpc InverseKinematics(InverseKinematicsReq) returns (InverseKinematicsRes) {}
  rpc ForwardKinematics(ForwardKinematicsReq) returns (ForwardKinematicsRes) {}
  rpc CheckAproachRetractValid(CheckAproachRetractValidReq) returns (CheckAproachRetractValidRes) {}
  rpc GetPalletPointList(GetPalletPointListReq) returns (GetPalletPointListRes) {}
  rpc CalculateRelativePose(CalculateRelativePoseReq) returns (CalculateRelativePoseRes) {}
  rpc CalculateCurrentPoseRel(CalculateCurrentPoseRelReq) returns (CalculateCurrentPoseRelRes) {}
  rpc PingFromConty(Empty) returns (Empty) {}

  // ---------------------------------------------------------------- //
  // TeleOperation
  // ---------------------------------------------------------------- //
  rpc GetTeleOpDevice(Empty) returns (TeleOpDevice) {}        // 프레임워크에 기록된 디바이스 받아옴, 없으면 리턴 값의 type = NONE
  rpc GetTeleOpState(Empty) returns (TeleOpState) {}          // 현재 텔레옵 상태 획득 -> 해당 모드 종료 외 버튼 상시 비활성화

  rpc ConnectTeleOpDevice(TeleOpDevice) returns (Response) {} // 텔레옵 장치 연결 (실패 시 code!=0, 메시지 리턴)
  rpc DisConnectTeleOpDevice(Empty) returns (Response) {}     // 텔레옵 장치 연결 끊기 (연결 정보는 남음, 실패 시 code!=0, 메시지 리턴)
  rpc ReadTeleOpInput(Empty) returns (TeleP) {}               // 교시 장치 입력 값 확인

  rpc StartTeleOp(TeleOpState) returns (Response) {}          // 칼리브레이션/기록/재생/RAW 모드 시작 (실패 시 code!=0, 메시지 리턴)
  rpc StopTeleOp(Empty) returns (Response) {}                 // 칼리브레이션/기록/재생/RAW 모드 종료 (실패 시 code!=0, 메시지 리턴)

  rpc SetPlayRate(TelePlayRate) returns (Response) {}         // 재생 속도 설정 (0~1)
  rpc GetPlayRate(Empty) returns (TelePlayRate) {}            // 재생 속도 확인 (0~1)

  rpc GetTeleFileList(Empty) returns (TeleOpFileList) {}      // 저장된 모션 리스트 획득
  rpc SaveTeleMotion(TeleFileReq) returns (Response) {}       // 현재 모션 파일로 저장
  rpc LoadTeleMotion(TeleFileReq) returns (Response) {}       // 모션 파일 불러오기
  rpc DeleteTeleMotion(TeleFileReq) returns (Response) {}       // 모션 파일 삭제

  rpc EnableTeleKey(State) returns (Response) {}
  rpc MoveTeleJ(MoveTeleJReq) returns (Response) {}       // RAW 메소드 활성 시 외부 입력
  rpc MoveTeleL(MoveTeleLReq) returns (Response) {}       // RAW 메소드 활성 시 외부 입력

  // ---------------------------------------------------------------- //
  // Sanding
  // ---------------------------------------------------------------- //
  rpc MoveLF(MoveLFReq) returns (Response) {}
  rpc GetTransformedFTSensorData(Empty) returns (TransformedFTSensorData) {}
}